### PR TITLE
**DRAFT** feat: Azure Service Bus Instrumentation

### PR DESCRIPTION
**Pull request by tippmar-nr**
Thank you for submitting a pull request.  Please review our [contributing guidelines](/CONTRIBUTING.md) and [code of conduct](https://opensource.newrelic.com/code-of-conduct/).

## Description

This PR adds instrumentation for Azure Service Bus, for applications that use the [`Azure.Messaging.ServiceBus` NuGet package](https://www.nuget.org/packages/Azure.Messaging.ServiceBus), v7.12.0 and later. 

The following methods are instrumented:
* `ServiceBusReceiver`: 
    * `ReceiveMessageAsync()` and `ReceiveMessagesAsync()`
    * `ReceiveDeferredMessageAsync()` and `ReceiveDeferredMessagesAsync()`
    * `PeekMessageAsync()` and `PeekMessagesAsync()`
    * `CompleteMessageAsync()`
    * `AbandonMessageAsync()`
    * `DeadLetterMessageAsync()`
    * `DeferMessageAsync()`
    * `RenewMessageLockAsync()`
* `ServiceBusSender`:
    * `SendMessageAsync()` and `SendMessagesAsync()`
    * `ScheduleMessageAsync()` and `ScheduleMessagesAsync()`
    * `CancelScheduledMessageAsync()` and `CancelScheduledMessagesAsync()`

Distributed tracing support is included, using the `ApplicationProperties` dictionary. 

# Author Checklist
- [x] Unit tests, Integration tests, and Unbounded tests completed
- [ ] Performance testing completed with satisfactory results (if required)

# Reviewer Checklist
- [ ] Perform code review
- [ ] Pull request was adequately tested (new/existing tests, performance tests)

### CODE REVIEW

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/Instrumentation.xml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+<!--
+Copyright 2020 New Relic Corporation. All rights reserved.
+SPDX-License-Identifier: Apache-2.0
+-->
+
+<extension xmlns="urn:newrelic-extension">
+  <instrumentation>
+
+    <tracerFactory name="AzureServiceBusReceiveWrapper">
+      <!--
```
Comment (jaffinito) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/Instrumentation.xml:11]: I like having the method signatures here.

File: tests/Agent/IntegrationTests/SharedApplications/Common/MFALatestPackages/MFALatestPackages.csproj [hunk 2]
```diff
@@ -8,6 +8,9 @@
     <PackageReference Include="AWSSDK.BedrockRuntime" Version="3.7.404.5" Condition="'$(TargetFramework)' == 'net481'" />
     <PackageReference Include="AWSSDK.BedrockRuntime" Version="3.7.404.5" Condition="'$(TargetFramework)' == 'net8.0'" />
     
+    <PackageReference Include="Azure.Messaging.ServiceBus" Version="7.18.2" Condition="'$(TargetFramework)' == 'net481'" />
```
Comment (jaffinito) [inline tests/Agent/IntegrationTests/SharedApplications/Common/MFALatestPackages/MFALatestPackages.csproj:11]: Is it possible to check another version of the library here, oldest supported and newest maybe?

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/AzureServiceBusReceiveWrapper.cs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    public override CanWrapResponse CanWrap(InstrumentedMethodInfo instrumentedMethodInfo)
+    {
+        var canWrap = instrumentedMethodInfo.RequestedWrapperName.Equals(nameof(AzureServiceBusReceiveWrapper));
+        return new CanWrapResponse(canWrap);
+    }
+
+    public override AfterWrappedMethodDelegate BeforeWrappedMethod(InstrumentedMethodCall instrumentedMethodCall, IAgent agent, ITransaction transaction)
+    {
+        dynamic serviceBusReceiver = instrumentedMethodCall.MethodCall.InvocationTarget;
+        string queueName = serviceBusReceiver.EntityPath; // marty-test-queue
```
Comment (jaffinito) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/AzureServiceBusReceiveWrapper.cs:28]: Like the examples, but should these be generic and not specific to your testing?

Comment (tippmar-nr) [reply src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/AzureServiceBusReceiveWrapper.cs]: Done

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/AzureServiceBusSendWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        string queueName = serviceBusReceiver.EntityPath; // some-queue-name
+        string fqns = serviceBusReceiver.FullyQualifiedNamespace; // some-service-bus-entity.servicebus.windows.net   
+
+        // determine message broker action based on method name
+        MessageBrokerAction action =
+            instrumentedMethodCall.MethodCall.Method.MethodName switch
+            {
+                "SendMessagesAsync" => MessageBrokerAction.Produce,
+                "ScheduleMessagesAsync" => MessageBrokerAction.Produce,
+                "CancelScheduledMessagesAsync" => MessageBrokerAction.Purge, // TODO is this correct ???
```
Comment (jaffinito) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/AzureServiceBusSendWrapper.cs:32]: This seems correct based on the description of the method.

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/AzureServiceBusWrapperBase.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+using NewRelic.Agent.Api;
+using NewRelic.Agent.Extensions.Providers.Wrapper;
+
+namespace NewRelic.Providers.Wrapper.AzureServiceBus
+{
+    public abstract class AzureServiceBusWrapperBase : IWrapper
+    {
+        protected const string BrokerVendorName = "AzureServiceBus";
+
+        public bool IsTransactionRequired => true; // only instrument service bus methods if we're already in a transaction
```
Comment (jaffinito) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/AzureServiceBusWrapperBase.cs:13]: This seems different compared with other MQ that we instrument.  Shouldn't the receive generate a transaction?

Comment (tippmar-nr) [reply src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/AzureServiceBusWrapperBase.cs:13]: Probably a topic for pairing on Tuesday. But this implementation is consistent with our MSMQ instrumentation. Unlike RabbitMQ, there's not an "eventing" layer that we could instrument that would wrap processing time on the client application.

Comment (tippmar-nr) [reply src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureServiceBus/AzureServiceBusWrapperBase.cs:13]: Looks like we could potentially instrument `ServiceBusProcessor` which does implement an event-based receive model where client code is called when a message is received. I'll take a quick look to see if it's worth the effort or not.

**Comment by tippmar-nr** [comment]
Closing pending further work

**Pull request closed without merging**